client/components/auth-form/auth-form.tsx

```diff
@@ -0,0 +1,77 @@
+import Input from '../form-controls/input';
+import Button from '../form-controls/button';
+import css from './auth-form.module.css';
+import { useAuthContext } from '../../hooks/auth-context';
+import { FormEvent, ReactElement, useState } from 'react';
+import { Link, Navigate } from 'react-router-dom';
+import { logIn, signUp } from '../../services/auth-service';
+import { UserToken } from '../../../common/api-types';
+
+type Method = 'log-in' | 'sign-up';
+type Props = { method: Method };
+
+export default function AuthForm({ method }: Props): ReactElement {
+  const [user, setUser] = useAuthContext();
+  const [errorMessage, setErrorMessage] = useState<string>();
+
+  if (user) {
+    return <Navigate to="/" replace={true} />;
+  }
+
+  const [actionPhrase, alternativePhrase, alternativePath, authFunction] = derivedValues[method];
+
+  async function handleSubmit(e: FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+    const formData = new FormData(e.target as HTMLFormElement);
+    const email: string = formData.get('email') as string;
+    const password: string = formData.get('password') as string;
+
+    const response = await authFunction(email, password);
+    console.log(response);
+    if (typeof response === 'string') {
+      setErrorMessage(response);
```

Comment:
So after following the thread here it looks like this error would read 'invalid email/password' (I'm definitely not 100% sure though). Have you considered more detailed error messages, like one for an invalid email and one for an invalid password?

Reply:
So, a bit of input on this one.
It's generally not a good idea to separate the invalid email/password messages. Not always, but it could indicate to people or programs with malicious intent that they have the right username, and now they just need to brute-force the password, or the other way around.
Detailed error messages are great... in every other context. But for this, I'd simply output "Invalid login credentials" or something similar.